### ISSUE TITLE
[Parity 5/16] Autograd completeness and torch.func - higher-order, forward-mode, vmap, jacobian/hessian, hooks, anomaly

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#214]
Part of epic #209.

## Summary

Our reverse-mode autograd is strong. What's missing is everything *else* PyTorch offers: higher-order gradients, forward-mode AD, the `torch.func` transforms (`vmap`, `grad`, `jvp`, `vjp`, `jacfwd`, `jacrev`, `hessian`, `functional_call`), saved-tensor hooks, anomaly detection, and `inference_mode` as a distinct state from `no_grad`.

These unlock per-sample gradients, meta-learning, ensembles, physics-informed models, implicit layers, Hessian-free optimization, probabilistic programming, and model debugging.

### Higher-order autograd
- `torch.autograd.grad(..., create_graph=True)` — grads of grads
- `autograd.functional.jacobian`, `hessian`, `hvp`, `vhp`, `jvp`, `vjp`
- Double-backward rules for every differentiable op

### Forward-mode AD
- Dual-number primal+tangent propagation
- `torch.autograd.forward_ad` dual-tensor API
- Forward-mode rules for every op (usually auto-derivable from reverse)

### `torch.func`
- `grad(fn)`, `grad_and_value(fn)`
- `jvp(fn, primals, tangents)`, `vjp(fn, *primals)`
- `jacfwd(fn)`, `jacrev(fn)`, `hessian(fn)`
- `vmap(fn, in_dims, out_dims)` — batched function execution
- `functional_call(module, params, args)` — stateless module application
- `stack_module_state`, `replace_all_batch_norm_modules_`

### Autograd plumbing
- `torch.no_grad` — disables grad tracking
- `torch.inference_mode` — disables grad tracking *and* version counter (allows in-place)
- `torch.autograd.set_detect_anomaly(True)` — backward NaN/Inf localization with full forward trace
- `saved_tensors_hooks` — custom pack/unpack for activations (CPU offload, quantization, compression)
- `torch.autograd.graph.save_on_cpu` — activation offload
- Backward hooks on tensors and modules

## Current state in the repo

- `GradientTape.cs`, `TapeEntry.cs`, `TapeEntryArena.cs`, `AutogradFunction.cs` — reverse mode works.
- `GradientTapeOptions.cs` — options present; need to verify `inference_mode` semantics.
- `SymbolicBackwardGraph.cs`, `CompiledBackwardGraph.cs` — excellent foundation for double-backward.
- `GradientCheckpointing.cs`, `ActivationCheckpoint.cs` — checkpointing works.
- No forward-mode AD. No `vmap`. No saved-tensor hooks. No anomaly detection. No `functional_call`.

## How we beat PyTorch

1. **`vmap` over compiled plans, not just eager**: PyTorch's `vmap` is a transform that runs op-by-op. We can insert `vmap` into `LazyTensorScope` so the fused+scheduled plan gets batched at the graph level. Target 2–5× on nested vmap calls.
2. **Forward-mode AD with SIMD dual numbers**: pack primal+tangent into a `Vector<T>` register so scalar ops double up. PyTorch forward-mode is serial per tensor.
3. **`functional_call` with zero-copy parameter override** via our existing `ParameterBuffer.cs` — PyTorch has to clone state dicts.
4. **Per-sample gradients via `vmap(grad(fn))` without the slow-path penalty**: route through `DataflowFusionPass` so the per-sample backward gets fused with its forward. Standard PyTorch trick but 3× faster.
5. **Algebraic simplification of Jacobians / Hessians**: we already have `AlgebraicSimplificationPass` — `hessian(f)` returns a compile-time-simplified Hessian. PyTorch computes numerically, we compute symbolically when possible.
6. **Saved-tensor hooks with built-in recipes**: ship `SaveOnCpu`, `SaveQuantized(int8/int4/NF4/FP4)`, `SaveZstd`, `SaveDelta` (diff from previous checkpoint). PyTorch only ships `save_on_cpu`.
7. **Anomaly mode that localizes to the original forward call** via tape step context (already have `TapeStepContext.cs`) with zero runtime cost when disabled — PyTorch pays a ~10% tax even with anomaly off.
8. **Inference mode that unlocks our in-place ops** and skips tape allocation entirely — PyTorch's `inference_mode` already does this; we need to verify `GradientTapeOptions` honors it.
9. **Backward hooks that understand compiled plans**: hook can be registered against a graph node, not just a tensor, so it survives fusion.

### Higher-order autograd
- [ ] `GradientTape.Grad(outputs, inputs, createGraph: true)`
- [ ] `Autograd.Functional.Jacobian(fn, inputs)` (forward + reverse-mode strategies)
- [ ] `Autograd.Functional.Hessian(fn, inputs)`
- [ ] `Autograd.Functional.Hvp(fn, inputs, vec)`, `Vhp`, `Jvp`, `Vjp`
- [ ] Double-backward rules audited for all differentiable ops

### Forward-mode AD
- [ ] `ForwardAD.Dual<T>` packed primal+tangent tensor
- [ ] `ForwardAD.Jvp(fn, primals, tangents)`
- [ ] Forward-mode op rules registered through `OpRegistry`

### `torch.func` equivalents (namespace: `Func` / static `TensorFunc`)
- [ ] `TensorFunc.Grad(fn)` — returns a function that computes gradient
- [ ] `TensorFunc.GradAndValue(fn)`
- [ ] `TensorFunc.Jvp(fn, primals, tangents)`
- [ ] `TensorFunc.Vjp(fn, primals)` → `(outputs, vjp_fn)`
- [ ] `TensorFunc.JacFwd(fn)`, `TensorFunc.JacRev(fn)`, `TensorFunc.Hessian(fn)`
- [ ] `TensorFunc.Vmap(fn, inDims, outDims)` — operating on compiled plans too
- [ ] `TensorFunc.FunctionalCall(module, parameters, args)` — stateless module apply

### Plumbing
- [ ] `InferenceMode` scope distinct from `NoGradMode`, allowing in-place without version counters
- [ ] `Autograd.SavedTensorHooks` (pack/unpack) + built-in recipes (SaveOnCpu, SaveQuantized, SaveZstd)
- [ ] `Autograd.AnomalyMode` with forward-op-site localization
- [ ] Backward hooks on tensors *and* graph nodes

## Acceptance criteria

- Double-backward tests pass for every op in `DifferentiableOps.cs`.
- Forward-mode + reverse-mode agree on every op (cross-validation test).
- `vmap(grad(fn))` for per-sample gradients matches a reference loop over the batch within tolerance.
- `jacobian`/`hessian` numerical parity vs `jax.jacobian` / `jax.hessian` reference (within tolerance).
- `functional_call` allows gradient-through-params flows used in MAML-style meta-learning.
- `InferenceMode` benchmark: no allocation for tape entries, in-place ops allowed, ≥10% speedup over `NoGradMode` for a standard inference loop.
- Saved-tensor `SaveQuantized(int8)` recipe demonstrates 4× activation memory reduction with ≤1% accuracy hit on a standard CNN training run.

## Implementation notes

- Forward-mode module under `src/AiDotNet.Tensors/Engines/Autodiff/ForwardAD/`.
- `torch.func` transforms under `src/AiDotNet.Tensors/Engines/Autodiff/Transforms/`.
- `vmap` needs to understand `LazyTensorScope` (so compile-mode works) and `TensorView` broadcasts.
- Saved-tensor hooks integrate with `TapeEntry.cs` and `TapeStepContext.cs`.
- Reuse `SymbolicBackwardGraph` for Jacobian/Hessian materialization.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state (file-level)

| File | Status |
|---|---|
| `Engines/Autodiff/GradientTape.cs` | Reverse mode ✓ |
| `Engines/Autodiff/TapeEntry.cs` + `TapeEntryArena.cs` | Tape storage ✓ |
| `Engines/Autodiff/AutogradFunction.cs` | Custom autograd.Function equivalent ✓ |
| `Engines/Autodiff/GradientTapeOptions.cs` | Options — `InferenceMode` semantics need audit |
| `Engines/Autodiff/SymbolicBackwardGraph.cs` | Excellent foundation for double-backward |
| `Engines/Autodiff/CompiledBackwardGraph.cs` | Compiled backward ✓ |
| `Engines/Autodiff/AlgebraicSimplificationPass.cs`, `BackwardCSEPass.cs`, `BackwardPruningPass.cs` | Unique to us |
| `Engines/Compilation/GradientCheckpointing.cs` | Checkpointing ✓ |
| `Engines/Autodiff/OpRegistry.cs`, `DifferentiableOps.cs`, `BackwardFunctions.cs` | Core ✓ |

**Missing:** forward-mode AD, `vmap`, all `torch.func` transforms, saved-tensor hooks, anomaly detection, `functional_call`, distinct `InferenceMode`.

## 2. Scope — full inventory

### Higher-order autograd
```
GradientTape.Grad(outputs, inputs, createGraph=true)
Autograd.Functional.{Jacobian, Hessian, Hvp, Vhp, Jvp, Vjp}(fn, inputs)
Double-backward rules audited across every op in DifferentiableOps
```

### Forward-mode AD
```
ForwardAD.Dual<T> { Primal; Tangent; }  — packed into Vector<T> where possible
ForwardAD.Jvp(fn, primals, tangents)
OpRegistry.RegisterForwardRule<TOp>(rule)
```

### `torch.func` transforms
```
TensorFunc.Grad(fn) → fn' that returns gradient
TensorFunc.GradAndValue(fn) → fn' returning (value, grad)
TensorFunc.Jvp(fn, primals, tangents) → (outputs, tangent_outputs)
TensorFunc.Vjp(fn, primals) → (outputs, vjp_fn)
TensorFunc.JacFwd(fn) → jacobian-fn via forward-mode
TensorFunc.JacRev(fn) → jacobian-fn via reverse-mode
TensorFunc.Hessian(fn) → hessian via JacFwd of JacRev
TensorFunc.Vmap(fn, inDims, outDims) → batched-fn
TensorFunc.FunctionalCall(module, parameters, args)
```

### Plumbing
```
InferenceMode scope — distinct from NoGradMode; allows in-place + skips version counter
Autograd.SavedTensorHooks (pack/unpack) + recipes: SaveOnCpu, SaveQuantized, SaveZstd, SaveDelta
Autograd.AnomalyMode with forward-op-site localization via TapeStepContext
Backward hooks on tensors + graph nodes (survive fusion)
Autograd.ReplaceAllBatchNormModules_ (for functional_call workflows)
```

**Total: ~20 public surfaces + backward-rule audits on every op (~90 ops in `DifferentiableOps`).**

## 3. Win vectors

1. **`vmap` over compiled plans**: insert `vmap` as a `LazyTensorScope` transform. Nested vmap becomes a single multi-batched plan. Target 2–5× vs PyTorch op-by-op vmap.
2. **SIMD dual numbers**: pack primal+tangent into `Vector<T>` register — scalar ops double up. PyTorch forward-mode is serial.
3. **`functional_call` zero-copy param override** via `ParameterBuffer.cs`. PyTorch clones state dicts.
4. **Per-sample gradients via `vmap(grad(fn))`**: route through `DataflowFusionPass` so per-sample backward fuses with forward. 3× over PyTorch's `_ExpandedWeights` trick.
5. **Algebraic simplification of Jacobians/Hessians**: `AlgebraicSimplificationPass` folds identity/zero/sparsity. `hessian(linear-function)` is 0 at compile time.
6. **Built-in saved-tensor recipes**: `SaveOnCpu`, `SaveQuantized(int8/int4/NF4/FP4)`, `SaveZstd`, `SaveDelta`. PyTorch ships only `save_on_cpu`.
7. **Zero-cost anomaly mode when disabled**: compile-time toggle — no runtime tax. PyTorch pays ~10% always-on.
8. **`InferenceMode`** unlocks our in-place ops + skips tape allocation entirely. Target ≥10% over `NoGradMode`.
9. **Backward hooks on graph nodes** — survive fusion, unlike PyTorch's tensor-only hooks.
10. **Double-backward parity across every op**: audit is systematic — inventory every op in `DifferentiableOps.cs` and verify second-order correctness.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| Double-backward | 100% ops in `DifferentiableOps.cs` pass second-order gradcheck |
| Forward vs reverse cross-validation | Every op: forward-mode grad matches reverse-mode grad within tolerance |
| `vmap(grad(fn))` per-sample grads | Match reference-loop output within tolerance; ≥3× faster via fusion |
| Nested `vmap` (batched batched) | ≥2× PyTorch equivalent |
| `jacobian`/`hessian` | Parity vs `jax.jacobian`/`jax.hessian` |
| `functional_call` | Enables MAML-style meta-learning end-to-end |
| `InferenceMode` | ≥10% speedup + zero tape alloc vs `NoGradMode` on ResNet-50 inference |
| `SaveQuantized(int8)` saved-tensor | 4× activation memory reduction, ≤1% accuracy on standard CNN train |
| Anomaly mode — disabled cost | ≤0.1% overhead (compile-time toggle) |
| All 6 GPU backends | `vmap`/forward-mode work across every backend (autograd is backend-agnostic) |

## 5. File inventory

**New (~50):**
- `src/AiDotNet.Tensors/Engines/Autodiff/ForwardAD/` — Dual, DualOps, JvpExecutor (~8)
- `src/AiDotNet.Tensors/Engines/Autodiff/Transforms/` — Vmap, VmapExecutor, Grad, Jvp, Vjp, JacFwd, JacRev, Hessian, FunctionalCall (~12)
- `src/AiDotNet.Tensors/Engines/Autodiff/Hooks/` — SavedTensorHooks + recipes (SaveOnCpu, SaveQuantized, SaveZstd, SaveDelta) (~6)
- `src/AiDotNet.Tensors/Engines/Autodiff/Anomaly/` — AnomalyMode, NaNDetector (~3)
- `src/AiDotNet.Tensors/Engines/Autodiff/Modes/` — InferenceModeScope, NoGradModeScope refactor (~2)
- `src/AiDotNet.Tensors/Engines/Autodiff/Functional/` — Autograd.Functional entry points (~5)
- `src/AiDotNet.Tensors/Engines/Autodiff/Rules/` — forward-mode rule registrations per op (~50 rule files split by category)

**Modified (~12):** `OpRegistry.cs`, `DifferentiableOps.cs`, `BackwardFunctions.cs`, `GradientTape.cs`, `GradientTapeOptions.cs`, `TapeEntry.cs`, `TapeStepContext.cs`, `SymbolicBackwardGraph.cs`, `LazyTensorScope.cs`, `LazyNode.cs`, `AlgebraicSimplificationPass.cs`, `DataflowFusionPass.cs`.

**Tests (~40):** double-backward per-op tests, forward↔reverse cross-validation, vmap parity, jacobian/hessian parity vs JAX, functional_call tests, SavedTensorHooks recipes, AnomalyMode tests, InferenceMode benchmark-gated tests, backward-hook-survives-fusion tests.

**Benchmarks (~4):** `VmapGradBenchmarks`, `ForwardModeVsReverseBenchmarks`, `SavedTensorQuantizedBenchmarks`, `InferenceModeBenchmarks`.

## 6. Test matrix

| Axis | Values |
|---|---|
| double-backward | Every op in `DifferentiableOps.cs` × 3 shapes × fp32/fp64 |
| forward vs reverse | Same — cross-validate outputs |
| vmap | in_dims ∈ {0, 1, -1, None}; composed with grad; nested |
| functional_call | MAML test: grad-through-params |
| SavedTensorHooks | SaveQuantized + round-trip + accuracy on CNN train |
| InferenceMode | zero tape alloc assert; in-place ops allowed assert |
| anomaly | NaN injection → correct forward-site localization |

## 7. Acceptance gate

1. Every op in `DifferentiableOps.cs` passes double-backward gradcheck.
2. Every op has a forward-mode rule registered.
3. `vmap(grad(fn))` batched per-sample gradients match ref-loop.
4. `jacobian/hessian` parity vs JAX.
5. `functional_call` enables working MAML test.
6. `InferenceMode` ≥10% over `NoGradMode` on ResNet-50.
7. `SaveQuantized(int8)` 4× memory reduction, ≤1% accuracy.
8. Anomaly disabled cost ≤0.1%.
9. All 6 backends work (autograd is backend-agnostic).

## 8. Timeline — ~5 weeks

W1 Forward-mode AD core (Dual, DualOps, forward-rule registration) + double-backward audit on 40 ops.
W2 Remaining double-backward audit + symbolic-simplification coverage.
W3 `torch.func` transforms (grad/jvp/vjp/jacfwd/jacrev/hessian).
W4 `vmap` over compiled plans + `functional_call`.
W5 SavedTensorHooks + AnomalyMode + InferenceMode + benchmarks + PR.

## 9. Dependencies

- **Blocks**: #223 (parametrizations use forward-mode jvp), #225 (AOTAutograd joint-graph needs vmap + jacfwd), partial dep of every differentiable-op PR landing after it.
- **Blocked by**: #210 (movement ops for vmap's dim manipulation), partially #212 (so jacfwd/jacrev can exercise FFT ops in tests).

## 10. Risks

- **vmap composability** — every new op must respect batched dims correctly. Audit burden.
- **Forward-mode rule correctness** — can auto-derive from reverse in most cases; verify via cross-validation tests.
- **`functional_call` state management** — must not silently mutate state.

---

*Part of epic #209. Plan ratified 2026-04-19.*
